pkg/vault/contracts/Vault.sol

```diff
@@ -268,29 +267,19 @@ contract Vault is IVaultMain, VaultCommon, Proxy {
     function _loadSwapState(
         SwapParams memory params,
         PoolData memory poolData
-    ) private view returns (SwapState memory state) {
-        // Use the storage map only for translating token addresses to indices. Raw balances can be read from poolData.
-        EnumerableMap.IERC20ToBytes32Map storage poolBalances = _poolTokenBalances[params.pool];
-
-        // EnumerableMap stores indices *plus one* to use the zero index as a sentinel value for non-existence.
-        uint256 indexIn = poolBalances.unchecked_indexOf(params.tokenIn);
-        uint256 indexOut = poolBalances.unchecked_indexOf(params.tokenOut);
+    ) private pure returns (SwapState memory state) {
+        int indexIn = _findTokenIndex(poolData.tokens, params.tokenIn);
+        int indexOut = _findTokenIndex(poolData.tokens, params.tokenOut);
 
         // If either are zero, revert because the token wasn't registered to this pool.
-        if (indexIn == 0 || indexOut == 0) {
+        if (indexIn == -1 || indexOut == -1) {
             // We require the pool to be initialized, which means it's also registered.
             // This can only happen if the tokens are not registered.
             revert TokenNotRegistered();
         }
 
-        // Convert to regular 0-based indices now, since we've established the tokens are valid.
-        unchecked {
-            indexIn -= 1;
-            indexOut -= 1;
-        }
-
-        state.indexIn = indexIn;
-        state.indexOut = indexOut;
+        state.indexIn = uint(indexIn);
```

Comment:
```suggestion
        state.indexIn = uint256(indexIn);
```
Here and elsewhere; we don't use the abbreviation. I do like not having the 0 vs. -1 conversion